process/block/shardblockProposal.go

```diff
@@ -212,9 +212,141 @@ func (sp *shardProcessor) VerifyBlockProposal(
 	// 	return nil, nil, err
 	// }
 
+	// TODO: the verification needs a cutoff handler as well - e.g stop after the block including the execution results for
+	// the configured cutoff round/nonce/epoch
+	// errCutoff := sp.blockProcessingCutoffHandler.HandleProcessErrorCutoff(header)
+
 	return nil
 }
 
+func getHaveTimeForProposal(startTime time.Time, maxDuration time.Duration) func() time.Duration {
+	timeOut := startTime.Add(maxDuration)
+	haveTime := func() time.Duration {
+		return time.Until(timeOut)
+	}
+	return haveTime
+}
+
+// ProcessBlockProposal processes the proposed block. It returns nil if all ok or the specific error
+func (sp *shardProcessor) ProcessBlockProposal(
+	headerHandler data.HeaderHandler,
+	bodyHandler data.BodyHandler,
+) (data.BaseExecutionResultHandler, error) {
+	if check.IfNil(headerHandler) {
+		return nil, process.ErrNilBlockHeader
+	}
+	if check.IfNil(bodyHandler) {
+		return nil, process.ErrNilBlockBody
+	}
+	if !headerHandler.IsHeaderV3() {
+		return nil, process.ErrInvalidHeader
+	}
+
+	sp.processStatusHandler.SetBusy("shardProcessor.ProcessBlock")
+	defer sp.processStatusHandler.SetIdle()
+
+	sp.roundNotifier.CheckRound(headerHandler)
+	sp.epochNotifier.CheckEpoch(headerHandler)
+	sp.requestHandler.SetEpoch(headerHandler.GetEpoch())
```

Comment:
perhaps move these after `err := sp.createBlockStarted()`? although this is how they used to be on the old shardBlock implementation